backend/store/cache/cache.go

```diff
@@ -0,0 +1,258 @@
+// store on creation.
+func New(ctx context.Context, client *clientv3.Client, resource corev2.Resource, synthesize bool) (*Resource, error) {
+	// Get the type of the resource and create a slice type of []type
+	typeOfResource := reflect.TypeOf(resource)
+	sliceOfResource := reflect.SliceOf(typeOfResource)
+	// Create a pointer to our slice type and then set the slice value
+	ptr := reflect.New(sliceOfResource)
+	ptr.Elem().Set(reflect.MakeSlice(sliceOfResource, 0, 0))
+
+	// Get a keybuilderFunc for this resource
+	keyBuilderFunc := func(ctx context.Context, name string) string {
+		return store.NewKeyBuilder(resource.StorePrefix()).WithContext(ctx).Build("")
+	}
+
+	err := etcd.List(ctx, client, keyBuilderFunc, ptr.Interface(), &store.SelectionPredicate{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating ResourceCacher: %s", err)
+	}
+
+	results := ptr.Elem()
+	resources := make([]corev2.Resource, results.Len())
+	for i := 0; i < results.Len(); i++ {
+		r, ok := results.Index(i).Interface().(corev2.Resource)
+		if !ok {
+			logger.Errorf("%T is not core2.Resource", results.Index(i).Interface())
+			continue
+		}
+		resources[i] = r
+	}
+
+	mapCache := make(map[key]Value, len(resources))
+	for _, resource := range resources {
+		mapCache[getCacheKey(resource)] = getCacheValue(resource, synthesize)
+	}
+
+	cache := &Resource{
+		sliceCache: MakeSliceCache(resources, synthesize),
+		mapCache:   mapCache,
+		watcher:    etcd.GetResourceWatcher(ctx, client, keyBuilderFunc(ctx, ""), typeOfResource),
+		synthesize: synthesize,
+	}
+	go cache.start(ctx)
+	return cache, nil
+}
+
+// NewFromResources creates a new resources cache using the given resources.
+// This function should only be used for testing purpose; it provides a way to
+// inject resources directly into the cache without an actual store
+func NewFromResources(resources []corev2.Resource, synthesize bool) *Resource {
+	return &Resource{
+		cacheMu:    sync.Mutex{},
+		sliceCache: MakeSliceCache(resources, synthesize),
+	}
+}
+
+// Watch allows cache users to get notified when the cache has new values.
+// When the context is canceled, the channel will be closed.
+func (r *Resource) Watch(ctx context.Context) <-chan struct{} {
+	watcher := cacheWatcher{
+		ctx: ctx,
+		ch:  make(chan struct{}, 1),
+	}
+	r.watchersMu.Lock()
+	r.watchers = append(r.watchers, watcher)
+	r.watchersMu.Unlock()
+	return watcher.ch
+}
+
+func (r *Resource) notifyWatchers() {
+	r.watchersMu.Lock()
+	defer r.watchersMu.Unlock()
+	deletes := map[int]struct{}{}
+	for i, watcher := range r.watchers {
+		if err := watcher.ctx.Err(); err != nil {
+			deletes[i] = struct{}{}
+			continue
+		}
+		select {
+		case watcher.ch <- struct{}{}:
+		default:
+			// if there is already a notification in the buffer, don't send
+			// another.
+		}
+	}
+	newWatchers := make([]cacheWatcher, 0, len(r.watchers))
+	for i, w := range r.watchers {
+		if _, ok := deletes[i]; !ok {
```

Comment:
Can you explain what `deletes` is tracking here. Unless I'm misunderstanding something, `deletes` is only populated with `struct{}{}` on L175.

Reply:
I believe Eric originally wrote that bit but my understanding is that we track the watchers in `r.watchers`.

Before notifying them, we make sure they were are still active here: https://github.com/sensu/sensu-go/blob/a1458e05a2df19d6e11befbf6a72aa2716055f93/backend/store/cache/cache.go#L168

If they were stopped, we add the specific watcher to the `deletes` slice: https://github.com/sensu/sensu-go/blob/a1458e05a2df19d6e11befbf6a72aa2716055f93/backend/store/cache/cache.go#L169

Which is later used here, to build a new list of active watchers: https://github.com/sensu/sensu-go/blob/a1458e05a2df19d6e11befbf6a72aa2716055f93/backend/store/cache/cache.go#L181

Makes sense?